trace_decoder/src/imp.rs

```diff
@@ -0,0 +1,641 @@
+            } = crate::type1::frontend(instructions)?;
+            (state, storage, code.into_iter().map(Into::into).collect())
+        }
+    })
+}
+
+/// Break `txns` into batches of length `hint`, prioritising creating at least
+/// two batches.
+///
+/// [`None`] represents a dummy transaction that should not increment the
+/// transaction index.
+fn batch(txns: Vec<TxnInfo>, hint: usize) -> Vec<Vec<Option<TxnInfo>>> {
+    let hint = cmp::max(hint, 1);
+    let mut txns = txns.into_iter().map(Some).collect::<Vec<_>>();
+    let n_batches = txns.iter().chunks(hint).into_iter().count();
+    match (txns.len(), n_batches) {
+        // enough
+        (_, 2..) => txns
+            .into_iter()
+            .chunks(hint)
+            .into_iter()
+            .map(FromIterator::from_iter)
+            .collect(),
+        // not enough batches at `hint`, but enough real transactions
+        (2.., ..2) => {
+            let second = txns.split_off(txns.len() / 2);
+            vec![txns, second]
+        }
+        // add padding
+        (0 | 1, _) => txns
+            .into_iter()
+            .pad_using(2, |_ix| None)
+            .map(|it| vec![it])
+            .collect(),
+    }
+}
+
+#[test]
+fn test_batch() {
+    #[track_caller]
+    fn do_test(n: usize, hint: usize, exp: impl IntoIterator<Item = usize>) {
+        itertools::assert_equal(
+            exp,
+            batch(vec![TxnInfo::default(); n], hint)
+                .iter()
+                .map(Vec::len),
+        )
+    }
+
+    do_test(0, 0, [1, 1]); // pad2
+    do_test(1, 0, [1, 1]); // pad1
+    do_test(2, 0, [1, 1]); // exact
+    do_test(3, 0, [1, 1, 1]);
+    do_test(3, 1, [1, 1, 1]);
+    do_test(3, 2, [2, 1]); // leftover after hint
+    do_test(3, 3, [1, 2]); // big hint
+}
+
+#[derive(Debug)]
+struct Batch<StateTrieT> {
+    pub first_txn_ix: usize,
+    pub gas_used: u64,
+    /// See [`GenerationInputs::contract_code`].
+    pub contract_code: BTreeSet<Vec<u8>>,
+    /// For each transaction in batch, in order.
+    pub byte_code: Vec<NonEmpty<Vec<u8>>>,
+
+    pub before: IntraBlockTries<StateTrieT>,
+    pub after: TrieRoots,
+
+    /// Empty for all but the final batch
+    pub withdrawals: Vec<(Address, U256)>,
+}
+
+/// [`evm_arithmetization::generation::TrieInputs`],
+/// generic over state trie representation.
+#[derive(Debug)]
+struct IntraBlockTries<StateTrieT> {
+    pub state: StateTrieT,
+    pub storage: BTreeMap<H256, StorageTrie>,
+    pub transaction: TransactionTrie,
+    pub receipt: ReceiptTrie,
+}
+
+/// Does the main work mentioned in the [module documentation](mod@self).
+fn middle<StateTrieT: StateTrie + Clone>(
+    // state at the beginning of the block
+    mut state_trie: StateTrieT,
+    // storage at the beginning of the block
+    mut storage_tries: BTreeMap<H256, StorageTrie>,
+    // None represents a dummy transaction that should not increment the transaction index
+    // all batches SHOULD not be empty
```

Comment:
> // all batches SHOULD not be empty

Maybe clarify something like `but may contain dummy TxnInfo` to clarify for the padding logic?

Reply:
I don't follow - in this code there is no dummy TxnInfo - dummies are always `None`